test/cluster/basic_test.go

```diff
@@ -290,3 +291,319 @@ func TestNodeRecoveryLeader(t *testing.T) {
+		require.NoError(t, err)
+		require.NotNil(t, rcpt)
+		require.True(t, txID != "")
+		require.True(t, len(rcpt.GetHeader().GetValidationInfo()) > 0)
+		require.Equal(t, types.Flag_VALID, rcpt.Header.ValidationInfo[rcpt.TxIndex].Flag)
+		t.Logf("tx submitted: %s, %+v", txID, rcpt)
+	}
+
+	require.NoError(t, c.Restart())
+
+	//find the new leader
+	newLeader := -1
+	require.Eventually(t, func() bool {
+		newLeader = c.AgreedLeader(t, 0, 1, 2)
+		return newLeader >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	//make sure all 3 nodes are active
+	require.Eventually(t, func() bool {
+		clusterStatusResEnv, err := c.Servers[newLeader].QueryClusterStatus(t)
+		return err == nil && clusterStatusResEnv != nil && len(clusterStatusResEnv.GetResponse().GetActive()) == 3
+	}, 30*time.Second, 100*time.Millisecond)
+
+	require.Eventually(t, func() bool {
+		return c.AgreedHeight(t, 3, 0, 1, 2)
+	}, 30*time.Second, 100*time.Millisecond)
+
+	for _, key := range keys {
+		require.Eventually(t, func() bool {
+			dataEnv, err := c.Servers[newLeader].QueryData(t, worldstate.DefaultDBName, key)
+			if dataEnv != nil && dataEnv.GetResponse().GetValue() != nil {
+				dataVal := dataEnv.GetResponse().GetValue()
+				require.Equal(t, dataVal, []byte(key+"-data"))
+				t.Logf("data: %+v", dataVal)
+			}
+			return dataEnv != nil && dataEnv.GetResponse().GetValue() != nil && err == nil
+		}, 30*time.Second, 100*time.Millisecond)
+
+	}
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - wait for one to be the leader.
+// - stop the leader.
+// - wait for new server to be the leader.
+// - the leader submit a tx
+// - repeat until each server was a leader at least once.
+// - make sure each node submitted a valid tx as a leader
+func TestAllNodesGetLeadership(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	leaders := [3]bool{false, false, false}
+
+	currentLeader := -1
+	require.Eventually(t, func() bool {
+		currentLeader = c.AgreedLeader(t, 0, 1, 2)
+		return currentLeader >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	oldLeader := -1
+	for !leaders[0] || !leaders[1] || !leaders[2] {
+		leaders[currentLeader] = true
+		txID, rcpt, err := c.Servers[currentLeader].WriteDataTx(t, worldstate.DefaultDBName, strconv.Itoa(currentLeader), []byte{uint8(currentLeader)})
```

Comment:
I am not sure `currentLeader` is a good candidate to the key, as a node might be elected twice, and therefore overwrite the same key. maybe a loop counter? see comment bellow as well.